### DISCUSSION TITLE
Load model via JSON String / direct input

### DISCUSSION THREAD

**Discussion by xL0b0** [alexandercerutti/passkit-generator#223] [Q&A] [answered]
Hello, is it possible to load a model and all the required properties into the pass payload, instead of through the „model“ parameter which reads a file, directly through an input into the generation payload?

**Comment by alexandercerutti** [comment, marked as answer]
Hi @xL0b0, thanks for using passkit-generator!

`PKPass` constructor accepts a list of file-buffer, as you can see in both examples and [documentation](https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#creating-a-pass).

https://github.com/alexandercerutti/passkit-generator/blob/bab4117cb029b00a50077262187fcc2ad9412357/examples/cloudflare-worker/src/index.ts#L44-L67

If you don't mean this, do you mind clarify further what you mean?

Let me know!

**Comment by xL0b0** [comment]
That's kinda what I wanted to do, this was my approach: 

```
    createPass({ parsedPass, index }) {
        return new Promise(async (resolve, reject) => {
            try {
                const pass = await PKPass.from({
                    model: parsedPass, // Pass the parsed JSON object directly, not as a string
                    certificates: {
                        wwdr: fs.readFileSync(`${this.documentPath}/Passkit/Certificates/wwdr.pem`),
                        signerCert: fs.readFileSync(`${this.documentPath}/Passkit/Certificates/signerCert.pem`),
                        signerKey: fs.readFileSync(`${this.documentPath}/Passkit/Certificates/signerKey.pem`),
                        signerKeyPassphrase: this.signerKeyPassphrase
                    }
                });

                const passBuffer = await pass.getAsBuffer();

                fs.writeFileSync(`./Passes/finished-${index}.pkpass`, passBuffer);
```

So I have a function that gets the completely parsed pass already from another function that fills in all the values into the JSON and then this function just writes the pass. So not reading them from a file, but passing it into the function dynamically.

**Reply by alexandercerutti** [reply]
Yeah, this way (you can find it as "buffer model" in the readme) was made on purpose for different modes, so that awaiting is not required.

Passing `pass.json` as a file in the buffer model, is a valid way to achieve things.

Furthermore, `model` accepts only a string, not an object, so I guess it was failing.

**Reply by alexandercerutti** [reply]
P.s. `getAsBuffer` is not async, it does not return a promise, so using `await` does not bring any plus to you.

**Reply by alexandercerutti** [reply]
P.s.2: if you could leave a ⭐ on the project, that would be awesome! 🚀

**Comment by xL0b0** [comment]
Tyvm, I will shortly give this a try and leave a star!

So in my case, I want to make an eventTicket, would that object also simply go into here then: 

```
const pass = new PKPass( 
 	{ 
 		"icon.png": Buffer.from(icon), 
 		"icon@2x.png": Buffer.from(icon2x), 
 		"footer.png": Buffer.from(footer), 
 		"footer@2x.png": Buffer.from(footer2x), 
 		"background@2x.png": Buffer.from(background2x), 
 	}, 
 	{ 
 		signerCert: env.SIGNER_CERT, 
 		signerKey: env.SIGNER_KEY, 
 		signerKeyPassphrase: env.SIGNER_PASSPHRASE, 
 		wwdr: env.WWDR, 
 	}, 
 	{ 
 		description: "Example Pass generated through a cloudflare worker", 
 		serialNumber: "81592CQ7838", 
 		passTypeIdentifier: "pass.com.passkitgenerator", 
 		teamIdentifier: "F53WB8AE67", 
 		organizationName: "Apple Inc.", 
 		foregroundColor: "rgb(255, 255, 255)", 
 		backgroundColor: "rgb(60, 65, 76)", 
eventTicket: { all the value here?}
 	}, 
 );
```

**Reply by alexandercerutti** [reply]
It is mostly correct.

However, `eventTicket` in this case is that `PKPass` considers a `type`, so you either set it with the [correct setter](https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#type-setter--getter) or you import it through your `pass.json` along with `icon.png` and the other files.

**Comment by xL0b0** [comment]
And right now I also have in my pass.json all my images as reference, in the images key

```
"images": {
                        "icon@2x.png": `./images/icon@2x.png`,
                        "logo@2x.png": `./images/logo@2x.png`,
                        "strip@2x.png": `./images/strip@2x.png`
                    },
```

Shall I leave this in still, but buffer all images too, like you did in your code example?

**Reply by alexandercerutti** [reply]
I suggest you to not call it `pass.json`, as `pass.json` is a specific file name in passes, with a specific format. What you are showing me here, instead, is something you did and not a standard format from Apple. So I suggest you renaming it in a different way, so you avoid confusion.

> but buffer all images too, like you did in your code example?

I don't understand what you are asking

**Comment by xL0b0** [comment]
I just want to set all variables for my event ticket pass via the arguments in the constructor of PKPass without using a pass.json file 😅

**Reply by alexandercerutti** [reply]
You can't. Different fields have different requirements in terms of validation.
What refrains you from using the setters methods and properties?

**Reply by alexandercerutti** [reply]
And, by the way, you can create your `pass.json` dynamically before calling the `new PKPass` and pass to it as a buffer instead as an object

**Reply by alexandercerutti** [reply]
Something like

```typescript
const passJson: PassProps = {
    // ... all your data in a valid format
};

const pkpassInstance = new PKPass({
    "pass.json": Buffer.from(JSON.stringify(passJson)),
    "icon.png": ...
}, {
    ...certificatesAndKeys
}, {
    ...otherDynamicProps
});
```

**Comment by xL0b0** [comment]
I just don't understand really what to now write into the otherDynamicProps since I got the entire data for the pass already in pass.json that I'm referencing in the first part of the arguments? Can you help me on that please?

**Reply by alexandercerutti** [reply]
Oh you can omit them if you have everything static. I mean, that object is meant for overriding some properties.

Still, I don't understand why you want to create you `pass.json` instead of using `otherDynamicProps` and setters/getters

**Comment by xL0b0** [comment]
Okay, my point is just simply here that I want to create a new Pass and instead of having to create files locally, like pass.json I just want to pass in my pass that I update with dynamic values, into the arguments of new PKPass, so I can then create a pass.

**Comment by xL0b0** [comment]
Okay, I made it, finally understood the issue!

**Reply by alexandercerutti** [reply]
What was it exactly?
